typing/typecore.ml

```diff
@@ -4685,6 +4728,121 @@ and type_function
+        let pck_ty = match pack.ctyp_desc with
+          | Ttyp_package pack -> pack
+          | _ -> assert false
+        in
+        let path = pck_ty.pack_path in
+        let fl = match get_desc pack.ctyp_type with
+            Tpackage (_, fl) -> fl
+          | _ -> assert false
+        in
+        (path, fl)
+      in
+      let (id_expected_typ_opt, (path, fl)) =
+        match split_function_mty env ty_expected
+                ~arg_label ~first ~in_function, pack with
+        | None, None ->
+          raise (Error (pparam_loc, env, Cannot_infer_signature))
+        | None, Some pack ->
+            None, type_pack pack
+        | Some (id, (path', fl'), ety), Some pack ->
+            let path, fl = type_pack pack in
+            begin try
+              unify env
+                (newty (Tfunctor (arg_label, id, (path, fl), newvar())))
+                (newty (Tfunctor (arg_label, id, (path', fl'), newvar())))
+            with Unify trace ->
+                raise (Error(loc, env, Expr_type_clash(trace, None, None)))
+            end;
+            (Some (id, ety), (path, fl))
+        | Some (id, (path', fl'), ety), None ->
+            if !Clflags.principal
+                && get_level ty_expected < Btype.generic_level
+            then Location.prerr_warning loc
+                  (not_principal "this module unpacking");
+            (Some (id, ety), (path', fl'))
+      in
+      !check_closed_package ~loc:name.loc ~env
+                           ~typ:(newty (Tpackage (path, fl))) fl;
+      let mty = !Ctype.modtype_of_package env pparam_loc path fl in
+      let pv_uid = Uid.mk ~current_unit:(Env.get_unit_name ()) in
+      let arg_md = {
+        md_type = mty;
+        md_attributes = [];
+        md_loc = pparam_loc;
+        md_uid = pv_uid;
+      } in
+      let (res_ty, params, body, newtypes, contains_gadt), s_ident =
+        with_local_level begin fun () ->
+          let s_ident =
+            Ident.create_scoped ~scope:(Ctype.get_current_level()) name.txt
+          in
+          let new_env = Env.add_module_declaration ~check:true s_ident
+                              Mp_present arg_md env in
+          let expected_res = match id_expected_typ_opt with
+            | Some (id, ety) ->
+              let topt = instance_funct ~id_in:(Ident.of_unscoped id)
+                                        ~p_out:(Pident s_ident)
+                                        ~fixed:false ety in
+              begin match topt with Some t -> t | None -> ety end
+            | None -> newvar ()
+          in
+          type_function new_env rest body_constraint body
+              expected_res ~first:false ~in_function,
+          s_ident
+      end
+      in
+      let ident = Ident.create_unscoped name.txt in
+      let exp_type =
+        match instance_funct ~p_out:(Pident (Ident.of_unscoped ident))
+                             ~id_in:s_ident ~fixed:false res_ty
+        with
```

Comment:
You should explain that the decision is taken in `copy_sep`, according to the presence or absence of free variables.